WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift

```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+            }
+        }
+
+        do {
+            try productReviewsResultsController.performFetch()
+        } catch {
+            ServiceLocator.crashLogging.logError(error)
+        }
+    }
+
+    func updateProductsResultsControllerPredicate(with productIDs: [Int64]) {
+        let predicates = NSCompoundPredicate(andPredicateWithSubpredicates: [sitePredicate(),
+                                                                            NSPredicate(format: "productID IN %@", productIDs)])
+        productsResultsController.predicate = predicates
+    }
+}
+
+// MARK: - Private helpers
+private extension ReviewsDashboardCardViewModel {
+    @MainActor
+    func loadReviews() async throws -> [ProductReview] {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductReviewAction.synchronizeProductReviews(siteID: siteID,
+                                                                          pageNumber: 1,
+                                                                          pageSize: Constants.numberOfItems) { result in
+                continuation.resume(with: result)
+            })
+        }
+    }
+
+    /// Updates data
+    @MainActor
+    func updateReviewsResults() async {
+        let reviews = productReviewsResultsController.fetchedObjects.prefix(Constants.numberOfItems)
+        let productIDs = reviews.map { $0.productID }
+
+        if productIDs.isNotEmpty {
+            updateProductsResultsControllerPredicate(with: productIDs)
+            do {
+                try await fetchProducts(for: productIDs)
+            } catch {
+                ServiceLocator.crashLogging.logError(error)
+            }
+        }
+
+        data = reviews
+            .map { review in
+                let product = reviewProducts.first { $0.productID == review.productID }
+                // TODO: also fetch notification
+                return ReviewViewModel(review: review, product: product, notification: nil)
+            }
+    }
+
+    /// Get products from storage if available, if not then fetch remotely.
+    ///
+    @MainActor
+    private func fetchProducts(for productIDs: [Int64]) async throws {
+        try productsResultsController.performFetch()
+
+        // Check if all productIDs are available in storage
+        let allProductsAvailable = productIDs.allSatisfy { productID in
+            reviewProducts.contains { $0.productID == productID }
+        }
+
+        if !allProductsAvailable {
+            await loadReviewProducts(for: productIDs)
+        }
+    }
+
+    @MainActor
+    func loadReviewProducts(for reviewProductIDs: [Int64]) async {
+        syncingData = true
+        syncingError = nil
+
+        do {
+            // Ignoring the result from remote as we're using storage as the single source of truth
+            _ = try await retrieveProducts(for: reviewProductIDs)
+        } catch {
+            syncingError = error
```

Comment:
Suggestion: I believe that we can display the reviews even if retrieving products fails. Could we ignore the error and just display the reviews without product info? 

If we think about it further, we could even load the products silently in the background and update the UI. What do you think?

Reply:
This is a neat suggestion and I can try adding it on the next PR that deals with loading states. I do notice that it can take a while waiting for products and notifications to be fetched, and showing partial result first can be more useful to users. 👍🏼